### PR TITLE
[MPS] port softmax to metal kernels

### PR DESCRIPTION
**Pull request by Isalia20**
PR is a bit big but to not get any visible slowdowns and good speedups, these kernels are needed.
Happy/easy path is last dim softmax which is usually the case in models.

Only losses are within 3 microseconds which I think is acceptable given the wins for 1x152k (final softmax in LLM over large vocab) is ~2x faster. Added tests as well to test out each specialized kernel (even though softmax test already existed but that might not have tested all of the kernels)

| label | dtype | shape | old (microseconds) | new (microseconds) | speedup |
|---|---|---|---:|---:|---|
| lm_4x32k | bf16 | 4x32000 | 19.2 | 21.5 | 0.89x |
| lm_4x32k | fp32 | 4x32000 | 20.9 | 22.8 | 0.92x |
| lm_4x32k | fp16 | 4x32000 | 19.8 | 21.6 | 0.92x |
| lm_4x152k | fp32 | 4x152064 | 49.0 | 51.0 | 0.96x |
| attn_1x32x2048 | fp32 | 1x32x2048x2048 | 3954.7 | 3941.9 | 1.00x |
| attn_1x32x1024 | fp32 | 1x32x1024x1024 | 1008.7 | 993.1 | 1.02x |
| attn_1x16x512 | fp32 | 1x16x512x512 | 130.7 | 128.6 | 1.02x |
| lm_4x152k | fp16 | 4x152064 | 44.3 | 43.6 | 1.02x |
| lm_4x152k | bf16 | 4x152064 | 45.1 | 43.4 | 1.04x |
| attn_1x32x2048 | fp16 | 1x32x2048x2048 | 2089.5 | 1977.6 | 1.06x |
| attn_1x32x2048 | bf16 | 1x32x2048x2048 | 2107.2 | 1984.5 | 1.06x |
| lm_1x50k | fp16 | 1x50257 | 47.1 | 43.4 | 1.09x |
| lm_1x50k | bf16 | 1x50257 | 47.0 | 41.9 | 1.12x |
| attn_1x32x1024 | fp16 | 1x32x1024x1024 | 561.9 | 501.4 | 1.12x |
| attn_1x32x1024 | bf16 | 1x32x1024x1024 | 572.4 | 505.1 | 1.13x |
| attn_1x16x512 | fp16 | 1x16x512x512 | 95.8 | 76.9 | 1.25x |
| lm_1x128k | fp16 | 1x128256 | 63.6 | 51.0 | 1.25x |
| attn_1x16x512 | bf16 | 1x16x512x512 | 95.7 | 65.5 | 1.46x |
| lm_1x32k | fp16 | 1x32000 | 64.0 | 40.9 | 1.57x |
| attn_1x8x128 | bf16 | 1x8x128x128 | 16.8 | 10.5 | 1.60x |
| attn_1x8x128 | fp16 | 1x8x128x128 | 19.0 | 11.4 | 1.67x |
| lm_1x32k | fp32 | 1x32000 | 69.8 | 41.0 | 1.70x |
| lm_1x50k | fp32 | 1x50257 | 79.7 | 46.3 | 1.72x |
| lm_1x128k | fp32 | 1x128256 | 91.3 | 52.5 | 1.74x |
| lm_1x32k | bf16 | 1x32000 | 61.5 | 35.1 | 1.75x |
| attn_1x8x128 | fp32 | 1x8x128x128 | 19.1 | 10.8 | 1.76x |
| lm_1x152k | bf16 | 1x152064 | 46.8 | 23.6 | 1.98x |
| lm_1x152k | fp16 | 1x152064 | 52.9 | 23.7 | 2.23x |
| lm_1x128k | bf16 | 1x128256 | 53.6 | 22.6 | 2.37x |
| lm_1x152k | fp32 | 1x152064 | 59.2 | 23.6 | 2.51x |

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/SoftMax.mm [hunk 1]
```diff
@@ -13,122 +13,271 @@
...
 
-  using namespace mps;
-  using CachedGraph = MPSUnaryCachedGraph;
-  MPSStream* stream = getCurrentMPSStream();
+    case V::SingleRow:
+    case V::Looped:
+    case V::General: {
+      std::string name;
+      if (choice.variant == V::SingleRow) {
+        name = "softmax_single_pass_" + type_str + "_" + std::to_string(choice.n_reads);
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/SoftMax.mm]: might cleaner with fmtlib?

Comment (Isalia20) [reply aten/src/ATen/native/mps/operations/SoftMax.mm]: done, thanks!

**Comment by Isalia20** [comment]
Closing this before this one is landed:
https://github.com/pytorch/pytorch/pull/181503
Didn't know there was PR up, will put up another one for non last dim once that's merged
